src/Hedgehog/Plutus/Tx.hs

```diff
@@ -0,0 +1,600 @@
+posDif = Plutus.unionWith (fmap (max 0) . (-))
+
+confirmBalanced :: TxContext -> Tx 'Unbalanced -> Maybe (Tx 'Balanced)
+confirmBalanced context tx =
+  (getBalance context tx >>= guard . (== mempty)) $> coerce tx
+
+data Spend = Spend
+  { spendUtxos :: Vector Plutus.TxOutRef
+  , excess :: Plutus.Value
+  }
+
+{- | Try to satisfy a value from available 'non-special' UTxOs matching the provided predicate.
+
+A UTxO is 'non-special' iff:
+
+* It is locked by a PubKey
+
+* It does not hold a datum
+
+The refs need not be held by the same PubKey.
+-}
+spendWhere ::
+  Model.Mock ->
+  Plutus.Value ->
+  (Plutus.TxOut -> Bool) ->
+  Maybe (Hedgehog.Gen Spend)
+spendWhere mock val pred = do
+  let utxos = Map.toList $ Map.filter (liftM2 (&&) nonSpecial pred) $ Model.mockUtxos mock
+  guard $ not $ null utxos
+  detSpend <- toRes $ foldl go (val, Spend Vector.empty mempty) utxos
+  -- create a deterministic spend to test that balance is possible
+  Just $ do
+    newUtxos <- Gen.shuffle utxos
+    let spend = toRes $ foldl go (val, Spend Vector.empty mempty) newUtxos
+    case spend of
+      Nothing -> pure detSpend
+      -- if the shuffled balance somehow fails
+      -- just return the deterministic spend
+      Just s -> pure s
+  where
+    nonSpecial :: Plutus.TxOut -> Bool
+    nonSpecial txout =
+      let
+        cred = Plutus.addressCredential $ Plutus.txOutAddress txout
+        noDatum = Plutus.txOutDatum txout == Plutus.NoOutputDatum
+        isPubKey = case cred of
+          Plutus.PubKeyCredential _ -> True
+          _ -> False
+       in
+        isPubKey && noDatum
+
+    toRes :: (Value, Spend) -> Maybe Spend
+    toRes (rem, s) = guard (rem == mempty) $> s
+
+    go ::
+      (Value, Spend) ->
+      (Plutus.TxOutRef, Plutus.TxOut) ->
+      (Value, Spend)
+    go (remaining, s@(Spend refs change)) (outRef, out)
+      | remaining == mempty = (remaining, s)
+      | intersect /= mempty =
+          (remaining', Spend (Vector.cons outRef refs) change')
+      | otherwise = (remaining, s)
+      where
+        refVal = Plutus.txOutValue out
+        intersect = Plutus.unionWith min remaining refVal
+        remaining' = posDif remaining intersect
+        change' = change <> posDif refVal intersect
+
+{- | Generate a @plutus-simple-model@ 'Model.Tx' from a @hedgehog-plutus-simple@
+ 'Tx'. This should automatically add neccesary scripts and signatures.
+-}
+toSimpleModelTx :: TxContext -> Tx bal -> Model.Tx
+toSimpleModelTx
+  TxContext
+    { interestingScripts
+    , mockchain
+    , datums
+    }
+  Tx
+    { txInputs
+    , txOutputs
+    , txMint
+    , txFee
+    , txValidRange
+    , txExtraSignatures
+    } =
+    Model.Tx
+      { Model.tx'extra = mempty
+      , Model.tx'plutus =
+          mempty
+            { Fork.txInputs = Set.map convertTxIn txInputs
+            , Fork.txCollateral = mempty
+            , -- TODO our tx type should probably support reference inputs
```

Comment:
My latest thinking is that reference inputs should be removed _entirely_ from `Tx`, along with `ScriptSource` and the `InScript` indirection. scripts can't see their sources AFAIK, and we only care about script validity, not script size, so this is only a complication.

Reply:
Validators can definitely distinguish normal inputs from reference inputs.
https://github.com/Plutonomicon/plutarch-plutus/blob/01a67f56b2bf428538e92ed9ada0ce88d90ab636/Plutarch/Api/V2/Contexts.hs#L54